// roots/lib/api/bail.js
// Generated by CoffeeScript 1.10.0
(function() {
  var RootsError,
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty;

  RootsError = (function(superClass) {
    extend(RootsError, superClass);

    function RootsError(name1, message1, ext1, code1) {
      this.name = name1;
      this.message = message1;
      this.ext = ext1;
      this.code = code1;
      Error.call(this);
      Error.captureStackTrace(this, this.constructor);
    }

    return RootsError;

  })(Error);

  module.exports = function(code, message, ext) {
    var name;
    switch (code) {
      case 125:
        name = "Malformed Extension";
        break;
      case 126:
        name = "Malformed Write Hook Output";
    }
    throw new RootsError(name, message, ext, code);
  };

}).call(this);
